client/heartbeatstop.go

```diff
@@ -0,0 +1,149 @@
+package client
+
+import (
+	"fmt"
+	"time"
+
+	hclog "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/nomad/client/state"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+type heartbeatStop struct {
+	lastOk      time.Time
+	allocs      map[string]time.Duration
+	allocHookCh chan *structs.Allocation
+	getRunner   func(string) (AllocRunner, error)
+	logger      hclog.InterceptLogger
+	state       state.StateDB
+}
+
+func newHeartbeatStop(
+	state state.StateDB,
+	getRunner func(string) (AllocRunner, error),
+	logger hclog.InterceptLogger) *heartbeatStop {
+
+	h := &heartbeatStop{
+		allocs:      make(map[string]time.Duration),
+		allocHookCh: make(chan *structs.Allocation),
+		getRunner:   getRunner,
+		logger:      logger,
+		state:       state,
+	}
+
+	if state != nil {
+		lastOk, err := state.GetLastHeartbeatOk()
+		if err == nil && lastOk != nil {
+			h.lastOk = *lastOk
+		}
+	}
+
+	return h
+}
+
+// allocHook is called after (re)storing a new AllocRunner in the client. It registers the
+// allocation to be stopped if the taskgroup is configured appropriately
+func (h *heartbeatStop) allocHook(alloc *structs.Allocation) {
+	tg := allocTaskGroup(alloc)
+	if tg.StopAfterClientDisconnect != nil {
+		h.allocHookCh <- alloc
+	}
+}
+
+// shouldStop is called on a restored alloc to determine if lastOk is sufficiently in the
+// past that it should be prevented from restarting
+func (h *heartbeatStop) shouldStop(alloc *structs.Allocation) bool {
+	tg := allocTaskGroup(alloc)
+	if tg.StopAfterClientDisconnect != nil {
+		now := time.Now()
+		return now.After(h.lastOk.Add(*tg.StopAfterClientDisconnect))
+	}
+	return false
+}
+
+// watch is a loop that checks for allocations that should be stopped. It also manages the
+// registration of allocs to be stopped in a single thread.
+func (h *heartbeatStop) watch() {
+	// If we never manage to successfully contact the server, we want to stop our allocs
+	// after duration + start time
+	h.lastOk = time.Now()
+	stop := make(chan string, 1)
+	var now time.Time
+	var interval time.Duration
+	checkAllocs := false
+
+	for {
+		// minimize the interval
+		interval = 5 * time.Second
+		for _, t := range h.allocs {
+			if t < interval {
+				interval = t
+			}
+		}
+
+		checkAllocs = false
+		timeout := time.After(interval)
+
+		select {
+		case allocID := <-stop:
+			if err := h.stopAlloc(allocID); err != nil {
+				h.logger.Warn("stopping alloc %s on heartbeat timeout failed: %v", allocID, err)
+				continue
+			}
+			delete(h.allocs, allocID)
+
+		case alloc := <-h.allocHookCh:
+			tg := allocTaskGroup(alloc)
+			if tg.StopAfterClientDisconnect != nil {
+				h.allocs[alloc.ID] = *tg.StopAfterClientDisconnect
+			}
+
+		case <-timeout:
+			checkAllocs = true
+		}
+
+		if !checkAllocs {
+			continue
+		}
+
+		now = time.Now()
+		for allocID, d := range h.allocs {
+			if now.After(h.lastOk.Add(d)) {
```

Comment:
I re-read through the golang docs on [monotonic clocks](https://golang.org/pkg/time/#hdr-Monotonic_Clocks) a couple times, and I'm concerned about the way we persist the last OK to the state store. My understanding is that a monotonic clock value won't have any meaning outside of a single process, so the monotonic clock value is discarded when we serialize it. I _think_ that means that after a restart we have only wall clock time, in which case if a client is restarted on either side of a DST change*, could this result in the client throwing away all its tasks?

(* I've seen the inside of a multi-$B startup who had all their AWS servers set to Pacific Time, so this isn't as far-fetched as one might think! 😆)

Reply:
We probably do this sort of thing lots of places, but just retry later and it's all ok. Here it's kind of a one-way trip.

Reply:
It does mean that we won't have a monotonic component on either side of a reboot, the constructors for parsing a timestamp from any representation produce them without a monotonic component. To take it in two pieces: I think that the only problem related specifically to monotonicity is that we could read a timestamp still in `Now`'s future, and I think that's safe. The DST thing is horrifying, so I've just changed it to serialize the unix timestamp. The comparison of that to `Now` should be ok.

Reply:
:+1: